pkg/frontend/openshiftcluster_preflightvalidation.go

```diff
@@ -0,0 +1,152 @@
+package frontend
+
+// Copyright (c) Microsoft Corporation.
+// Licensed under the Apache License 2.0.
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"strings"
+
+	"github.com/Azure/go-autorest/autorest/to"
+	"github.com/go-chi/chi/v5"
+	"github.com/sirupsen/logrus"
+
+	"github.com/Azure/ARO-RP/pkg/api"
+	"github.com/Azure/ARO-RP/pkg/env"
+	"github.com/Azure/ARO-RP/pkg/frontend/middleware"
+)
+
+// /subscriptions/{subscriptionId}/resourcegroups/{resourceGroupName}/providers/{resourceProviderNamespace}/deployments/{deploymentName}/preflight?api-version={api-version}
+func (f *frontend) preflightValidation(w http.ResponseWriter, r *http.Request) {
+	ctx := r.Context()
+	log := ctx.Value(middleware.ContextKeyLog).(*logrus.Entry)
+
+	var header http.Header
+	var b []byte
+
+	body := r.Context().Value(middleware.ContextKeyBody).([]byte)
+	originalPath := r.Context().Value(middleware.ContextKeyOriginalPath).(string)
+	referer := r.Header.Get("Referer")
+
+	subId := chi.URLParam(r, "subscriptionId")
+	resourceGroupName := chi.URLParam(r, "resourceGroupName")
+
+	resources, err := unmarshalRequest(body)
+	if err == nil {
+		resourceCount := len(resources.Resources)
+		ch := make(chan api.ValidationResult, resourceCount)
+		for _, raw := range resources.Resources {
+			// get typeMeta from the raw data
+			typeMeta := &api.ResourceTypeMeta{}
+			err = json.Unmarshal(raw, &typeMeta)
+			if err != nil {
+				// failing to parse the preflight body is not considered a validation failure. continue
+				log.Warningf("bad request. Failed to unmarshal ResourceTypeMeta: %s", err)
+				continue
+			}
+			if strings.EqualFold(typeMeta.Type, "Microsoft.RedHatOpenShift/openShiftClusters") {
+				path := PreflightResourceId(subId, resourceGroupName, typeMeta.Name)
+				ch <- f._preflightValidation(ctx, log, raw, path, originalPath, r.Method, referer, &header, f.apis[typeMeta.APIVersion].OpenShiftClusterConverter, f.apis[typeMeta.APIVersion].OpenShiftClusterStaticValidator)
+			}
+		}
+		close(ch)
+
+		for res := range ch {
+			// go through and return the first error
+			if res.Status == api.ValidationStatusFailed {
+				log.Warningf("preflight validation failed")
+				b, err = marshalValidationResult(&res)
+				reply(log, w, header, b, err)
+				return
+			}
+			b, err = marshalValidationResult(&res)
+		}
+		log.Info("preflight validation succeeded")
+	}
+	reply(log, w, header, b, err)
+}
+
+func (f *frontend) _preflightValidation(ctx context.Context, log *logrus.Entry, raw json.RawMessage, path, originalPath, method, referer string, header *http.Header, converter api.OpenShiftClusterConverter, staticValidator api.OpenShiftClusterStaticValidator) api.ValidationResult {
```

Comment:
i think you should remove unused parameters such as `referer` and `header`

Reply:
Removed unused parameters